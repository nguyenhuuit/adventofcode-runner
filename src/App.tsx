import { useStore } from 'zustand';

import React, { useEffect, useMemo } from 'react';

import { Box } from 'ink';

import { useExecuteAsStream } from '@hooks/useExecuteAsStream';
import { createExecutionStore } from '@hooks/useExecutionStore';
import { useHandleInput } from '@hooks/useHandleInput';
import { useInputFile } from '@hooks/useInputFile';
import { useSolutionFile } from '@hooks/useSolutionFile';
import { useYearInfo } from '@hooks/useYearInfo';

import FileSection from '@components/FileSection';
import HeaderSection from '@components/HeaderSection';
import ResultSection from '@components/ResultSection';

type Props = {
  promptInput: Required<PromtOptions> & { baseDir: string };
};

const App = ({ promptInput }: Props) => {
  const executionStore = useMemo(() => createExecutionStore(promptInput), []);

  const { year, day, language, part, output, answer, perfLog, loading } = useStore(executionStore);

  const { userName, star } = useYearInfo(executionStore);

  const { name: solutionFileName, size: solutionFileSize } = useSolutionFile(executionStore);

  const { name: inputFileName, size: inputFileSize } = useInputFile(executionStore);

  const executeSolution = useExecuteAsStream(executionStore);

  useEffect(() => {
    executeSolution();
  }, []);

  useHandleInput(executionStore);

  return (
    <>
      <HeaderSection
        year={year}
        day={day}
        part={part}
        language={language}
        userName={userName}
        star={star}
      />
      <Box height={1} />
      <FileSection
        solutionFileName={solutionFileName}
        solutionFileSize={solutionFileSize}
        inputFileName={inputFileName}
        inputFileSize={inputFileSize}
      />
      <Box height={1} />
      <ResultSection loading={loading} answer={answer} perfLog={perfLog} output={output} />
    </>
  );
};

export default App;
