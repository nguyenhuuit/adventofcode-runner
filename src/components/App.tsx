import React, { useEffect, useState } from 'react';
import { Box, Text, useApp, useInput } from 'ink';
import chokidar from 'chokidar';
import chalk from 'chalk';
import Spinner from './Spinner.js';
import { useSolutionFile } from '../hooks/useSolutionFile.js';
import { useInputFile } from '../hooks/useInputFile.js';
import { terminate } from '../utils/executors.js'
import { HELP_MESSAGE } from './constants.js';
import { useYearInfo } from '../hooks/useYearInfo.js';
import { useSubmit } from '../hooks/useSubmit.js';
import { useExecuteAsStream } from '../hooks/useExecuteAsStream.js';

const watcher = chokidar.watch([])

type Props = {
	state: AppState
}

const App = ({ state }: Props) => {
	const {exit} = useApp();

	const [tsUserName, setTsUserName] = useState(0)
	const { userName, star } = useYearInfo(state.year, tsUserName)

	const [inputMode, setInputMode] = useState(state.inputMode)
	const [part, setPart] = useState(state.part)
	const [output, setOutput] = useState(state.output);
	const [answer, setAnswer] = useState<any>('');
	const [perfLog, setPerfLog] = useState('');
	const [loading, setLoading] = useState(false);

	const [tsSolutionFile, setTsSolutionFile] = useState(0)
	const { name: solutionFileName, size: solutionFileSize } = useSolutionFile(state.year, state.day, part, state.language, tsSolutionFile);
	const [tsInputFile, setTsInputFile] = useState(0)
	const { name: inputFileName, size: inputFileSize } = useInputFile(state.year, state.day, inputMode, tsInputFile);

	const executeSolution = useExecuteAsStream({
		onOutput: (s: string) => setOutput(current => current ? current + s : s),
		onResult: (s: string) => setAnswer(s),
		onExecutionTime: (s: string) => setPerfLog(s),
		onStart: () => { setLoading(true); setOutput(""); },
		onExit: () => setLoading(false)
	})

	const submit = useSubmit(state.year, state.day, part, answer);

	const handleSubmit = async () => {
		setLoading(true)
		try {
			const { correct, waitingTime } = await submit();
			let msg = correct ? chalk.bold(chalk.greenBright('Right answer! 🤩')) : chalk.bold(chalk.red('Wrong answer! 🥹'))
			if (waitingTime) {
				msg += chalk.bold(chalk.redBright(` Waiting ${waitingTime} ⏳`))
			}
			setOutput(msg)
			setTsUserName(s => s + 1)
		} catch (err: any) {
			setOutput(chalk.bold(chalk.red(`Cannot submit answer: ${err}`)))
		} finally {
			setLoading(false)
		}
	}

	useEffect(() => {
		if (!solutionFileName) return;
		watcher.add(solutionFileName);
		watcher.removeAllListeners()
		watcher.on('change', async () => {
			const s = { year: state.year, day: state.day, part, inputMode, language: state.language }
			setTsSolutionFile(s => s + 1)
			executeSolution(s)
		});
		return () => {
			watcher.unwatch(solutionFileName)
		}
	}, [solutionFileName])
	useEffect(() => {
		if (!inputFileName) return;
		watcher.add(inputFileName);
		watcher.removeAllListeners()
		watcher.on('change', async () => {
			const s = { year: state.year, day: state.day, part, inputMode, language: state.language }
			setTsInputFile(s => s + 1)
			executeSolution(s)
		});
		return () => {
			watcher.unwatch(inputFileName)
		}
	}, [inputFileName])

	useInput(async (input, key) => {
		switch (input.toLowerCase()) {
			case 'q': {
				terminate();
				exit();
				process.exit();
			}
			case 'i': {
				setInputMode("input");
				const s = { year: state.year, day: state.day, part, inputMode: 'input', language: state.language }
				executeSolution(s)
				break;
			}
			case 's': {
				setInputMode("sample");
				const s = { year: state.year, day: state.day, part, inputMode: 'sample', language: state.language }
				executeSolution(s)
				break;
			}
			case 'c': setOutput(''); break;
			case '0':
			case '1':
			case '2':
			case '3':
			case '4':
			case '5':
			case '6':
			case '7':
			case '8':
			case '9': {
				setPart(input);
				const s = { year: state.year, day: state.day, part: input, inputMode, language: state.language }
				executeSolution(s)
				break;
			}
			case 'x': {
				if (loading) {
					terminate();
					setTimeout(() => setOutput('Terminated!'), 100)
				}
				break;
			}
			case 'h': {
				setOutput(HELP_MESSAGE);
				break;
			}
			case 'u': {
				handleSubmit();
				break;
			}
		}
		if (key.downArrow) {
			setInputMode("input")
			const s = { year: state.year, day: state.day, part, inputMode: 'input', language: state.language }
			executeSolution(s)
		}
		if (key.upArrow) {
			setInputMode("sample")
			const s = { year: state.year, day: state.day, part, inputMode: 'sample', language: state.language }
			executeSolution(s)
		}
		if (key.return) {
			const s = { year: state.year, day: state.day, part, inputMode, language: state.language }
			executeSolution(s)
		}
	})
	useEffect(() => {
		const s = { year: state.year, day: state.day, part, inputMode, language: state.language }
		executeSolution(s)
		return () => {}
	}, [])
	return (
		<>
			<Box>
				<Text>📆 Year:</Text>
				<Text italic bold color={"#FF8800"}> {state.year} </Text>
				<Text>Day:</Text>
				<Text italic bold color={"#FF8800"}> {state.day} </Text>
				<Text>Part:</Text>
				<Text italic bold color={"#FF8800"}> {part} </Text>
				<Text>| 📘 Language:</Text>
				<Text italic bold color={"#FF8800"}> {state.language} </Text>
				{userName && (
					<>
						<Text>| 👤 Username:</Text>
						<Text italic bold color={"#FF8800"}> {userName} </Text>
					</>
				)}
				{star && <Text italic bold color={"#FF8800"}>{star}⭐️ </Text>}
			</Box>
			<Box height={1}/>
			<Box width={100}>
				<Box flexDirection="column" width={70}>
					<Text>{`Solution file: ${chalk.bold(chalk.magentaBright(solutionFileName.slice(2)))} ${solutionFileSize} bytes`}</Text>
					<Text>{`   Input file: ${chalk.bold(chalk.magentaBright(inputFileName.slice(2)))} ${inputFileSize} bytes`}</Text>
				</Box>
			</Box>
			<Box height={1}/>
			<Box>
				<Text color="yellowBright"><Spinner type="earth" loading={loading}/></Text>
				<Text color="green">{`Result: `}</Text>
				<Text color="green" bold>{ loading ? <Spinner type="simpleDots" loading={true}/> : answer }</Text>
				{ (answer && !loading) && <Text>{`   ⏱ ${perfLog}`}</Text>}
			</Box>
			
			{output && (
				<>
					<Box height={1}/>
					<Box borderTop flexDirection="column">
						<Text>{output}</Text>
					</Box>
				</>
			)}
			
		</>
		
	);
}

export default App;
