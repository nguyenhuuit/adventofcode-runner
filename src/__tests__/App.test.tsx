import { render } from '@testing-library/react';

import React from 'react';

import App from '../App';

// Mock all hooks
jest.mock('@hooks/useExecuteAsStream', () => ({
  useExecuteAsStream: jest.fn(() => jest.fn()),
}));

jest.mock('@hooks/useHandleInput', () => ({
  useHandleInput: jest.fn(),
}));

jest.mock('@hooks/useInputFile', () => ({
  useInputFile: jest.fn(() => ({
    name: './input.txt',
    size: 123,
  })),
}));

jest.mock('@hooks/useSolutionFile', () => ({
  useSolutionFile: jest.fn(() => ({
    name: './solution.js',
    size: 456,
  })),
}));

jest.mock('@hooks/useYearInfo', () => ({
  useYearInfo: jest.fn(() => ({
    userName: 'TestUser',
    star: '42',
  })),
}));

// Mock child components
jest.mock('@components/FileSection', () => {
  return function MockFileSection(props: object) {
    return <div data-testid="file-section" {...props} />;
  };
});

jest.mock('@components/HeaderSection', () => {
  return function MockHeaderSection(props: object) {
    return <div data-testid="header-section" {...props} />;
  };
});

jest.mock('@components/ResultSection', () => {
  return function MockResultSection(props: object) {
    return <div data-testid="result-section" {...props} />;
  };
});

// Mock ink components
interface MockBoxProps {
  children: React.ReactNode;
  height?: number;
  [key: string]: unknown;
}

jest.mock('ink', () => ({
  Box: ({ children, ...props }: MockBoxProps) => (
    <div data-testid="box" {...props}>
      {children}
    </div>
  ),
}));

describe('App', () => {
  const defaultProps = {
    promptInput: {
      year: '2024',
      day: '1',
      part: '1',
      language: 'javascript',
      baseDir: './',
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with default props', () => {
    const { container } = render(<App {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with different prompt input', () => {
    const customProps = {
      promptInput: {
        year: '2023',
        day: '25',
        part: '2',
        language: 'python',
        baseDir: '/custom/path',
      },
    };

    const { container } = render(<App {...customProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should call useHandleInput hook', () => {
    const { useHandleInput } = require('@hooks/useHandleInput');
    render(<App {...defaultProps} />);

    expect(useHandleInput).toHaveBeenCalled();
  });

  it('should call useExecuteAsStream hook', () => {
    const { useExecuteAsStream } = require('@hooks/useExecuteAsStream');
    render(<App {...defaultProps} />);

    expect(useExecuteAsStream).toHaveBeenCalled();
  });

  it('should call useYearInfo hook', () => {
    const { useYearInfo } = require('@hooks/useYearInfo');
    render(<App {...defaultProps} />);

    expect(useYearInfo).toHaveBeenCalled();
  });

  it('should call useSolutionFile hook', () => {
    const { useSolutionFile } = require('@hooks/useSolutionFile');
    render(<App {...defaultProps} />);

    expect(useSolutionFile).toHaveBeenCalled();
  });

  it('should call useInputFile hook', () => {
    const { useInputFile } = require('@hooks/useInputFile');
    render(<App {...defaultProps} />);

    expect(useInputFile).toHaveBeenCalled();
  });
});
