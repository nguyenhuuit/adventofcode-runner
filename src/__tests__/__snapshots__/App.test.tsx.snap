// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly with default props 1`] = `
<div>
  <div
    data-testid="header-section"
    day="1"
    language="javascript"
    part="1"
    star="42"
    username="TestUser"
    year="2024"
  />
  <div
    data-testid="box"
    height="1"
  />
  <div
    data-testid="file-section"
    inputfilename="./input.txt"
    inputfilesize="123"
    solutionfilename="./solution.js"
    solutionfilesize="456"
  />
  <div
    data-testid="box"
    height="1"
  />
  <div
    answer=""
    data-testid="result-section"
    output=""
    perflog=""
  />
</div>
`;

exports[`App should render correctly with different prompt input 1`] = `
<div>
  <div
    data-testid="header-section"
    day="25"
    language="python"
    part="2"
    star="42"
    username="TestUser"
    year="2023"
  />
  <div
    data-testid="box"
    height="1"
  />
  <div
    data-testid="file-section"
    inputfilename="./input.txt"
    inputfilesize="123"
    solutionfilename="./solution.js"
    solutionfilesize="456"
  />
  <div
    data-testid="box"
    height="1"
  />
  <div
    answer=""
    data-testid="result-section"
    output=""
    perflog=""
  />
</div>
`;
